starboard/tools/testing/test_coverage.py

```diff
@@ -0,0 +1,186 @@
+  return [os.path.join(binaries_dir, binary) for binary in test_binaries]
+
+
+def _get_exclude_opts():
+  """Get list of exclude options."""
+  # Get a list of all directories in the src root.
+  all_dirs = next(os.walk(SRC_DIR))[1]
+  # TODO(oxv): Make configurable?
+  include_dirs = ['cobalt', 'starboard']
+  exclude_dirs = [d for d in all_dirs if d not in include_dirs]
+  return [
+      r'--ignore-filename-regex=.*_test\.h$',
+      r'--ignore-filename-regex=.*_test\.cc$',
+      r'--ignore-filename-regex=.*_test_fixture\.h$',
+      r'--ignore-filename-regex=.*_test_fixture\.cc$',
+      r'--ignore-filename-regex=.*_test_internal\.h$',
+      r'--ignore-filename-regex=.*_test_internal\.cc$',
+      r'--ignore-filename-regex=.*_test_util\.h$',
+      r'--ignore-filename-regex=.*_test_util\.cc$',
+      r'--ignore-filename-regex=.*_unittest\.h$',
+      r'--ignore-filename-regex=.*_unittest\.cc$',
+  ] + [f'--ignore-filename-regex={d}/.*' for d in exclude_dirs]
+
+
+def _generate_reports(target_paths, merged_data_path, report_dir, report_type):
+  """Generate reports based on the combined coverage data."""
+  toolchain_dir = build.GetClangBinPath(clang.GetClangSpecification())
+  llvm_cov = os.path.join(toolchain_dir, 'llvm-cov')
+
+  exclude_opts = _get_exclude_opts()
+  test_binary_opts = [target_paths[0]] + \
+    ['-object=' + target for target in target_paths[1:]]
+
+  if report_type in ['html', 'both']:
+    logging.info('Creating html coverage report')
+    show_cmd_list = [
+        llvm_cov, 'show', *exclude_opts, '-instr-profile=' + merged_data_path,
+        '-format=html', '-output-dir=' + os.path.join(report_dir, 'html'),
+        *test_binary_opts
+    ]
+    subprocess.check_call(show_cmd_list, text=True)
+
+  if report_type in ['gcov', 'both']:
+    logging.info('Creating gcov coverage report')
+    report_cmd_list = [
+        llvm_cov, 'show', *exclude_opts, '-instr-profile=' + merged_data_path,
+        '-format=text', *test_binary_opts
+    ]
+    with open(os.path.join(report_dir, 'report.txt'), 'wb') as out:
+      subprocess.check_call(report_cmd_list, stdout=out, text=True)
+
+
+def _merge_coverage_data(targets, coverage_dir):
+  """Create a combined coverage file from the raw reports."""
+  toolchain_dir = build.GetClangBinPath(clang.GetClangSpecification())
+  llvm_profdata = os.path.join(toolchain_dir, 'llvm-profdata')
+
+  raw_reports = _get_raw_reports(targets, coverage_dir)
+  if not raw_reports:
+    raise RuntimeError('No coverage data found in \'coverage_dir\'')
+
+  logging.info('Merging coverage files')
+  merged_data_path = os.path.join(coverage_dir, 'report.profdata')
+  merge_cmd_list = [
+      llvm_profdata, 'merge', '-sparse=true', '-o', merged_data_path
+  ] + raw_reports
+  subprocess.check_call(merge_cmd_list, text=True)
+
+  return merged_data_path
+
+
+def create_report(binary_dir, test_binaries, coverage_dir, report_type='both'):
```

Comment:
Should we add a success return value to this function so we can check for errors and report when we call it in test_runner.py?

Reply:
I don't think that the outcome of the coverage should affect the ultimate status of the test run.

Reply:
I just realized that the current implementation will do that. It raises exceptions currently so I think that wrapping it in try/catch in test_runner is probably the right thing to do.